// @cucumber/gherkin/dist/src/TokenExceptions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnexpectedEOFException = exports.UnexpectedTokenException = void 0;
const createLocation_1 = __importDefault(require("./createLocation"));
const Errors_1 = require("./Errors");
class UnexpectedTokenException extends Errors_1.GherkinException {
    static create(token, expectedTokenTypes) {
        const message = `expected: ${expectedTokenTypes.join(', ')}, got '${token.getTokenValue().trim()}'`;
        const location = tokenLocation(token);
        return this._create(message, location);
    }
}
exports.UnexpectedTokenException = UnexpectedTokenException;
class UnexpectedEOFException extends Errors_1.GherkinException {
    static create(token, expectedTokenTypes) {
        const message = `unexpected end of file, expected: ${expectedTokenTypes.join(', ')}`;
        const location = tokenLocation(token);
        return this._create(message, location);
    }
}
exports.UnexpectedEOFException = UnexpectedEOFException;
function tokenLocation(token) {
    return token.location && token.location.line && token.line && token.line.indent !== undefined
        ? createLocation_1.default({
            line: token.location.line,
            column: token.line.indent + 1,
        })
        : token.location;
}
